doc/src/manual/multi-threading.md

```diff
@@ -223,6 +223,70 @@ julia> a
 
 Note that [`Threads.@threads`](@ref) does not have an optional reduction parameter like [`@distributed`](@ref).
 
+### Using `@threads` without data races
+
+Taking the example of a naive sum
+
+```julia-repl
+julia> function sum_single(a)
+           s = 0
+           for i in a
+               s += i
+           end
+           s
+       end
+sum_single (generic function with 1 method)
+
+julia> sum_single(1:1_000_000)
+500000500000
+```
+
+Simply adding `@threads` exposes a data race with multiple threads reading and writing `s` at the same time.
+```julia-repl
+julia> function sum_multi_bad(a)
+           s = 0
+           Threads.@threads for i in a
+               s += i
+           end
+           s
+       end
+sum_multi_bad (generic function with 1 method)
+
+julia> sum_multi_bad(1:1_000_000)
+70140554652
+```
+
+Note that the result is not `500000500000` as it should be, and will most likely change each evaluation.
+
+To fix this, buffers that are specific to the task may be used to segment the sum into chunks that are race-free.
+
+```julia-repl
+julia> function sum_multi_good(a)
```

Comment:
I'm not an expert on multi-threaded parallelism, so I might just be wrong.

But this doesn't seem like it'll be particularly performant as the values of the `WeakKeyDict` won't be contiguous in memory, and so the reduction at the needs to gather these from likely different parts of the heap => slow reduction. In fact, when I tried running something like this on an example of mine it was incredibly slow; probably partially due to what I just mentioned, and partially due to the slowness of `WeakKeyDict` (and using something like `Dict` doesn't actually seem work).

An alternative is `Atomic` as mentioned below, or, even better, you can do a `Vector{Atomic{T}}` of some `buffer_length`. In the latter scenario you can then just randomly pick an index for each `Task`, and act on the corresponding `Atomic{T}` atomically, e.g. `atomic_add!(buffer[rand(1:buffer_size)], i)`. If the work within each of the tasks is fairly uniform, then this random picking of index to add to should result in fairly evenly congestion for the different buffers.

I'm sure there are much, much better ways of doing this, but the above is a quick-and-easy way to implement a much faster version of this kind of map-reduce approach using a buffer.